rfcs/git-sdk-requirements/RFC.md

```diff
@@ -0,0 +1,697 @@
+
+`add` = scans over added files, create entry in `.git` (even if many files added, should be instant)
+
+Most other git commands like `rebase` or `commit` won't do much else so all operations should be near instant.
+
+> go through the list of to be supported git commands and be more thorough in analysis of perf
+
+> ideally benchmark some commands
+
+We can also potentially delegate some heavy git operations to a web worker. To run some things in parallel and not block the main and only JS thread. Given the explanation presented in the previous section, this option should not be needed. However, offloading IO or computational work to web worker and waiting for response is possible if needed.
+
+#### Q: Does it make sense to run Isomoprphic Git and/or file system in a web worker?
+
+No, for above reason. It's not worth the complexity and there is no need for it.
+
+#### Q: How difficult would it be to add missing commands?
+
+Hard to predict but Isomorphic Git has some active contributors still. For example [abortMerge](https://github.com/isomorphic-git/isomorphic-git/pull/1744) was added recently.
+
+A `sparse-checkout` or `rebase` command would be taking that PR as template and making the logic work for respective command.
+
+By checking out how the code is done in quite a few git implementations out there. Some have already implemented it. In the worst case, you can read main [Git code](https://github.com/git/git) and figure out how those commands work from first principles and implement them.
+
+Paying attention that we are running in a browser context and not all details are needed, simplifies things a lot.
+
+#### Q: Will using Isomorphic Git answer to all stated goals above?
+
+✅ = already done
+🚧 = work required
+❎ = not possible
+
+Goal 1: Must run in the browser/on the client [High Confidence] ✅
+
+Goal 2: Lazy loading of files and git history [High Confidence] 🚧
+
+Is achievable in a few ways. But in current state would require sparse checkout to be implemented.
+
+Goal 3: Must be git compatible but not necessarily up to spec [Medium Confidence] ✅
+
+No issue here. Isomorphic Git is already git compatible, if we decide to extend it with new features we can have a fork of Isomorphic Git potentially or bring commands up to spec and merge them into Isomorphic Git repo.
+
+Goal 4: (Future) File-based auth [High Confidence | Server-related ] 🚧
+
+Is part of server so outside of Iso Git scope.
+
+> note: maybe wrong
+
+Goal 5: (Future) Support for large files out of the box [Medium Confidence] 🚧
+
+> note: need to research more to give good estimate how doable this is to add
+
+There is [open issue on this](https://github.com/isomorphic-git/isomorphic-git/issues/1375).
+
+Goal 6: (Future) Plugin system to support different files [Medium Confidence] 🚧
+
+Would be doable to add irrespective of whether libgit2 or Isomorphic Git is chosen.
```

Comment:
Please elaborate. 

I assume that plugins will be written in JS (that can call WASM under the hood). A JS implementation of git seems to serve us better because we can natively import JS plugins. Choosing libgit2 would likely complicate things.